### PR TITLE
Update github.com/ugorji/go/codec to latest version

### PR DESCRIPTION
**Pull request by zhsj**
New version moves the RawToString and StringToRaw options to
an embedded struct.

### CODE REVIEW

File: packer/rpc/server.go [hunk 1]
```diff
@@ -140,8 +140,12 @@ func (s *Server) Serve() {
 	defer stream.Close()
 
 	h := &codec.MsgpackHandle{
-		RawToString: true,
-		WriteExt:    true,
+		BasicHandle: codec.BasicHandle{
```
Comment (sylviamoss) [inline packer/rpc/server.go:143]: BasicHandle is deprecated, but it seems that they deprecated it 3 years ago 👀  https://github.com/ugorji/go/blame/master/codec/helper.go#L630

**Comment by sylviamoss** [comment]
I believe @SwampDragons will have more information about if we can update the lib or not.

**Comment by zhsj** [comment]
> 2020/07/14 10:03:02 packer-builder-docker plugin: Serving a plugin connection...
2020/07/14 10:03:02 packer-builder-docker plugin: Error detecting context: '' expected a map, got 'string' 

what's the template you use? I can build with packer-builder-docker.

<details>
<summary>log</summary>

```
zhsj@zhsj-debian:/tmp/t$ cat docker.json 
{
  "builders": [
    {
      "name": "ubuntu-docker",
      "type": "docker",
      "image": "ubuntu",
      "commit": true
    }
  ],
  "provisioners": [
    {
      "type": "shell",
      "inline": "echo 1 > /test"
    }
  ]
}
zhsj@zhsj-debian:/tmp/t$ PACKER_LOG=1 packer build docker.json 
2020/07/14 17:09:20 [INFO] Packer version: 1.6.1-dev [go1.14.4 linux amd64]
2020/07/14 17:09:20 Checking 'PACKER_CONFIG' for a config file path
2020/07/14 17:09:20 'PACKER_CONFIG' not set; checking the default config file path
2020/07/14 17:09:20 Attempting to open config file: /home/zhsj/.packerconfig
2020/07/14 17:09:20 [WARN] Config file doesn't exist: /home/zhsj/.packerconfig
2020/07/14 17:09:20 Setting cache directory: /tmp/t/packer_cache
2020/07/14 17:09:20 Creating plugin client for path: /home/zhsj/go/bin/packer
2020/07/14 17:09:20 Starting plugin: /home/zhsj/go/bin/packer []string{"/home/zhsj/go/bin/packer", "plugin", "packer-builder-docker"}
2020/07/14 17:09:20 Waiting for RPC address for: /home/zhsj/go/bin/packer
2020/07/14 17:09:20 packer-builder-docker plugin: [INFO] Packer version: 1.6.1-dev [go1.14.4 linux amd64]
2020/07/14 17:09:20 packer-builder-docker plugin: Checking 'PACKER_CONFIG' for a config file path
2020/07/14 17:09:20 packer-builder-docker plugin: 'PACKER_CONFIG' not set; checking the default config file path
2020/07/14 17:09:20 packer-builder-docker plugin: Attempting to open config file: /home/zhsj/.packerconfig
2020/07/14 17:09:20 packer-builder-docker plugin: [WARN] Config file doesn't exist: /home/zhsj/.packerconfig
2020/07/14 17:09:20 packer-builder-docker plugin: Setting cache directory: /tmp/t/packer_cache
2020/07/14 17:09:20 packer-builder-docker plugin: args: []string{"packer-builder-docker"}
2020/07/14 17:09:20 packer-builder-docker plugin: Plugin address: unix /tmp/packer-plugin365770819
2020/07/14 17:09:20 packer-builder-docker plugin: Waiting for connection...
2020/07/14 17:09:20 Received unix RPC address for /home/zhsj/go/bin/packer: addr is /tmp/packer-plugin365770819
2020/07/14 17:09:20 packer-builder-docker plugin: Serving a plugin connection...
2020/07/14 17:09:20 Creating plugin client for path: /home/zhsj/go/bin/packer
2020/07/14 17:09:20 Starting plugin: /home/zhsj/go/bin/packer []string{"/home/zhsj/go/bin/packer", "plugin", "packer-provisioner-shell"}
2020/07/14 17:09:20 Waiting for RPC address for: /home/zhsj/go/bin/packer
2020/07/14 17:09:20 packer-provisioner-shell plugin: [INFO] Packer version: 1.6.1-dev [go1.14.4 linux amd64]
2020/07/14 17:09:20 packer-provisioner-shell plugin: Checking 'PACKER_CONFIG' for a config file path
2020/07/14 17:09:20 packer-provisioner-shell plugin: 'PACKER_CONFIG' not set; checking the default config file path
2020/07/14 17:09:20 packer-provisioner-shell plugin: Attempting to open config file: /home/zhsj/.packerconfig
2020/07/14 17:09:20 packer-provisioner-shell plugin: [WARN] Config file doesn't exist: /home/zhsj/.packerconfig
2020/07/14 17:09:20 packer-provisioner-shell plugin: Setting cache directory: /tmp/t/packer_cache
2020/07/14 17:09:20 packer-provisioner-shell plugin: args: []string{"packer-provisioner-shell"}
2020/07/14 17:09:20 packer-provisioner-shell plugin: Plugin address: unix /tmp/packer-plugin532793842
2020/07/14 17:09:20 packer-provisioner-shell plugin: Waiting for connection...
2020/07/14 17:09:20 Received unix RPC address for /home/zhsj/go/bin/packer: addr is /tmp/packer-plugin532793842
2020/07/14 17:09:20 packer-provisioner-shell plugin: Serving a plugin connection...
2020/07/14 17:09:20 Preparing build: ubuntu-docker
2020/07/14 17:09:20 Build debug mode: false
2020/07/14 17:09:20 Force build: false
2020/07/14 17:09:20 On error: 
2020/07/14 17:09:20 Waiting on builds to complete...
2020/07/14 17:09:20 Starting build run: ubuntu-docker
2020/07/14 17:09:20 Running builder: docker
2020/07/14 17:09:20 [INFO] (telemetry) Starting builder docker
ubuntu-docker: output will be in this color.

2020/07/14 17:09:20 packer-builder-docker plugin: [DEBUG] Docker version: 19.0.0
2020/07/14 17:09:20 packer-builder-docker plugin: [DEBUG] Container will be committed
==> ubuntu-docker: Creating a temporary directory for sharing data...
2020/07/14 17:09:20 packer-builder-docker plugin: Set Packer temp dir to /home/zhsj/.packer.d/tmp805890246
==> ubuntu-docker: Pulling Docker image: ubuntu
2020/07/14 17:09:20 packer-builder-docker plugin: Executing: /usr/bin/docker [pull ubuntu]
    ubuntu-docker: Using default tag: latest
    ubuntu-docker: latest: Pulling from library/ubuntu
    ubuntu-docker: Digest: sha256:55cd38b70425947db71112eb5dddfa3aa3e3ce307754a3df2269069d2278ce47
    ubuntu-docker: Status: Image is up to date for ubuntu:latest
    ubuntu-docker: docker.io/library/ubuntu:latest
==> ubuntu-docker: Starting docker container...
    ubuntu-docker: Run command: docker run -v /home/zhsj/.packer.d/tmp805890246:/packer-files -d -i -t --entrypoint=/bin/sh -- ubuntu
2020/07/14 17:09:26 packer-builder-docker plugin: Starting container with args: [run -v /home/zhsj/.packer.d/tmp805890246:/packer-files -d -i -t --entrypoint=/bin/sh -- ubuntu]
2020/07/14 17:09:26 packer-builder-docker plugin: Waiting for container to finish starting
    ubuntu-docker: Container ID: f35033010373a4daacbc91d8457a1e6359239940a1586823b2627ad7cf8be65a
==> ubuntu-docker: Using docker communicator to connect: 172.17.0.2
2020/07/14 17:09:27 packer-builder-docker plugin: Running the provision hook
2020/07/14 17:09:27 [INFO] (telemetry) Starting provisioner shell
==> ubuntu-docker: Provisioning with shell script: /tmp/packer-shell735678633
2020/07/14 17:09:27 packer-provisioner-shell plugin: Opening /tmp/packer-shell735678633 for reading
2020/07/14 17:09:27 packer-provisioner-shell plugin: [INFO] 28 bytes written for 'uploadData'
2020/07/14 17:09:27 [INFO] 28 bytes written for 'uploadData'
2020/07/14 17:09:27 packer-builder-docker plugin: Copying to /tmp/script_7109.sh on container f35033010373a4daacbc91d8457a1e6359239940a1586823b2627ad7cf8be65a.
2020/07/14 17:09:27 packer-builder-docker plugin: Copied 28 bytes for /tmp/script_7109.sh
2020/07/14 17:09:28 packer-builder-docker plugin: Executing docker exec -i f35033010373a4daacbc91d8457a1e6359239940a1586823b2627ad7cf8be65a /bin/sh -c (chmod 0755 /tmp/script_7109.sh):
2020/07/14 17:09:28 packer-builder-docker plugin: [INFO] RPC endpoint: Communicator ended with: 0
2020/07/14 17:09:28 [INFO] RPC client: Communicator ended with: 0
2020/07/14 17:09:28 [INFO] RPC endpoint: Communicator ended with: 0
2020/07/14 17:09:28 packer-provisioner-shell plugin: [INFO] RPC client: Communicator ended with: 0
2020/07/14 17:09:28 packer-builder-docker plugin: Executing docker exec -i f35033010373a4daacbc91d8457a1e6359239940a1586823b2627ad7cf8be65a /bin/sh -c (chmod +x /tmp/script_7109.sh; PACKER_BUILDER_TYPE='docker' PACKER_BUILD_NAME='ubuntu-docker'  /tmp/script_7109.sh):
2020/07/14 17:09:28 packer-builder-docker plugin: [INFO] RPC endpoint: Communicator ended with: 0
2020/07/14 17:09:28 [INFO] 0 bytes written for 'stdout'
2020/07/14 17:09:28 [INFO] 0 bytes written for 'stderr'
2020/07/14 17:09:28 [INFO] RPC client: Communicator ended with: 0
2020/07/14 17:09:28 [INFO] RPC endpoint: Communicator ended with: 0
2020/07/14 17:09:28 packer-provisioner-shell plugin: [INFO] 0 bytes written for 'stdout'
2020/07/14 17:09:28 packer-provisioner-shell plugin: [INFO] 0 bytes written for 'stderr'
2020/07/14 17:09:28 packer-provisioner-shell plugin: [INFO] RPC client: Communicator ended with: 0
2020/07/14 17:09:28 packer-builder-docker plugin: Executing docker exec -i f35033010373a4daacbc91d8457a1e6359239940a1586823b2627ad7cf8be65a /bin/sh -c (rm -f /tmp/script_7109.sh):
2020/07/14 17:09:29 packer-builder-docker plugin: [INFO] RPC endpoint: Communicator ended with: 0
2020/07/14 17:09:29 [INFO] RPC client: Communicator ended with: 0
2020/07/14 17:09:29 [INFO] RPC endpoint: Communicator ended with: 0
2020/07/14 17:09:29 packer-provisioner-shell plugin: [INFO] RPC client: Communicator ended with: 0
2020/07/14 17:09:29 packer-builder-docker plugin: Executing docker exec -i f35033010373a4daacbc91d8457a1e6359239940a1586823b2627ad7cf8be65a /bin/sh -c (rm -f ):
2020/07/14 17:09:29 packer-builder-docker plugin: [INFO] RPC endpoint: Communicator ended with: 0
2020/07/14 17:09:29 [INFO] RPC client: Communicator ended with: 0
2020/07/14 17:09:29 [INFO] RPC endpoint: Communicator ended with: 0
2020/07/14 17:09:29 packer-provisioner-shell plugin: [INFO] RPC client: Communicator ended with: 0
2020/07/14 17:09:29 [INFO] (telemetry) ending shell
==> ubuntu-docker: Committing the container
2020/07/14 17:09:29 packer-builder-docker plugin: Committing container with args: [commit f35033010373a4daacbc91d8457a1e6359239940a1586823b2627ad7cf8be65a]
    ubuntu-docker: Image ID: sha256:1bbb6ca00a3b62610315c406106ead02a9d6ffd8382d1e3c6e58840016cb2b92
==> ubuntu-docker: Killing the container: f35033010373a4daacbc91d8457a1e6359239940a1586823b2627ad7cf8be65a
2020/07/14 17:09:30 [INFO] (telemetry) ending docker
==> Builds finished. The artifacts of successful builds are:
2020/07/14 17:09:30 machine readable: ubuntu-docker,artifact-count []string{"1"}
Build 'ubuntu-docker' finished.

==> Builds finished. The artifacts of successful builds are:
2020/07/14 17:09:30 machine readable: ubuntu-docker,artifact []string{"0", "builder-id", "packer.post-processor.docker-import"}
2020/07/14 17:09:30 machine readable: ubuntu-docker,artifact []string{"0", "id", "sha256:1bbb6ca00a3b62610315c406106ead02a9d6ffd8382d1e3c6e58840016cb2b92"}
2020/07/14 17:09:30 machine readable: ubuntu-docker,artifact []string{"0", "string", "Imported Docker image: sha256:1bbb6ca00a3b62610315c406106ead02a9d6ffd8382d1e3c6e58840016cb2b92"}
2020/07/14 17:09:30 machine readable: ubuntu-docker,artifact []string{"0", "files-count", "0"}
2020/07/14 17:09:30 machine readable: ubuntu-docker,artifact []string{"0", "end"}
2020/07/14 17:09:30 [INFO] (telemetry) Finalizing.
--> ubuntu-docker: Imported Docker image: sha256:1bbb6ca00a3b62610315c406106ead02a9d6ffd8382d1e3c6e58840016cb2b92
2020/07/14 17:09:31 waiting for all plugin processes to complete...
2020/07/14 17:09:31 /home/zhsj/go/bin/packer: plugin process exited
2020/07/14 17:09:31 /home/zhsj/go/bin/packer: plugin process exited
```

</details>

**Comment by sylviamoss** [comment]
I tested using HCL2 templates. The docker one is: 

```
variables {
    source_image= "centos:7.3.1611"
    username= "blahblahuser"
}

source "docker" "base_setup"{
    image = var.source_image
    commit = true
}

build {
    sources = ["docker.base_setup"]
    provisioner "shell" {
        inline = [ "useradd -m -s /bin/bash -G wheel var.username"]
    }
}
```

**Comment by SwampDragons** [comment]
What is the motivation for this update?

**Comment by zhsj** [comment]
Not really it needs to be updated. And I now find the `ugorji/go/codec` is hard to understand its behaviour for the new version. This patch now works find, but I don't understand why...

Anyway I'm fine to give up this patch :rofl:

**Comment by SwampDragons** [comment]
Thanks - I think this is probably a situation where we should hold off on updating just-because, but I'm going to let @azr weigh in.

**Comment by azr** [comment]
Hello there, I've dug up things and I think we should probably close this too.

I've been in 'our archives' about this and ugorji/go/codec has been breaking backwards compatibility a lot in the past so it is locked ( sometimes forked and locked ) in most Hashicorp products, so if we want to do this we might want to do it in a minor version ( where breaking changes are allowed ) since some packer plugins could stop working and need recompilation after this.

Then, as @sylviamoss stated this is breaking their builds for HCL2 config files so we would have to debug/fix this first before merging. I also tried to see if there were some improvements or bugfixes we could benefit from but I could not find a changelog file for the project.

Lastly, in the future we would like Packer to move over to using https://github.com/hashicorp/go-plugin which is going to be it's own thing and I think we should do that update then then if need be as we will have that context in mind then.

Many thanks for your time @zhsj for trying though, I'm going to go ahead and close this one since it seems like to me you only wanted to bump the version and PR it. 

That said if you want to spend the time addressing fixes for HCL2 here please feel free to re-open a PR with HCL2 working, we will probably merge for the next minor version.

**Pull request closed without merging**